vmm/task/src/youki.rs

```diff
@@ -0,0 +1,646 @@
+                .with_root_path(PathBuf::from(YOUKI_DIR))
+                .map_err(other_error!(e, "failed to set youki root path"))?
+                .with_console_socket(socket_path);
+
+            // if let Some(f) = container_io {
+            //     if let Some(fd) = f.stdin {
+            //         builder = builder.with_stdin(fd);
+            //     }
+            //     if let Some(fd) = f.stdout {
+            //         builder = builder.with_stdout(fd);
+            //     }
+            //     if let Some(fd) = f.stderr {
+            //         builder = builder.with_stderr(fd);
+            //     }
+            // }
+
+            builder
+                .as_tenant()
+                .with_detach(true)
+                .with_process(Some(&probe_path_clone))
+                .build()
+                .map_err(other_error!(
+                    e,
+                    "failed to exec process in youki container "
+                ))
+        })
+        .await
+        .map_err(other_error!(e, "failed to wait exec thread "))?;
+        tokio::fs::remove_file(&probe_path)
+            .await
+            .unwrap_or_default();
+        match exec_result {
+            Ok(pid) => {
+                copy_io_or_console(p, socket, pio, p.lifecycle.exit_signal.clone()).await?;
+                p.pid = pid.as_raw();
+                p.state = Status::RUNNING;
+            }
+            Err(e) => {
+                if let Some(s) = socket {
+                    s.clean().await;
+                }
+                return Err(other!("failed to start youki exec: {}", e));
+            }
+        }
+        Ok(())
+    }
+
+    async fn kill(
+        &self,
+        p: &mut ExecProcess,
+        signal: u32,
+        _all: bool,
+    ) -> containerd_shim::Result<()> {
+        if p.pid <= 0 {
+            Err(Error::FailedPreconditionError(
+                "process not created".to_string(),
+            ))
+        } else if p.exited_at.is_some() {
+            Err(Error::NotFoundError("process already finished".to_string()))
+        } else {
+            // TODO this is kill from nix crate, it is os specific, maybe have annotated with target os
+            kill(
+                Pid::from_raw(p.pid),
+                nix::sys::signal::Signal::try_from(signal as i32).unwrap(),
+            )
+            .map_err(Into::into)
+        }
+    }
+
+    async fn delete(&self, _p: &mut ExecProcess) -> containerd_shim::Result<()> {
+        self.exit_signal.signal();
+        Ok(())
+    }
+
+    async fn update(&self, _p: &mut ExecProcess, _resources: &LinuxResources) -> Result<()> {
+        Err(Error::Unimplemented("exec update".to_string()))
+    }
+
+    async fn stats(&self, _p: &ExecProcess) -> Result<Metrics> {
+        Err(Error::Unimplemented("exec stats".to_string()))
+    }
+
+    async fn ps(&self, _p: &ExecProcess) -> Result<Vec<ProcessInfo>> {
+        Err(Error::Unimplemented("exec ps".to_string()))
```

Comment:
I saw `InitProcess` has implemented `ps` trait, so `ExecProcess` maybe same.

Reply:
It is also not implemented in the container.rs, the ps method will only called with the init process.